---
title: "Sulk Notebook2"
output:
  pdf_document: default
  html_document:
    df_print: paged
---

Define basic parameters

```{r Parameters}
#minimum.read.number
mrn <- 6

#Raw Database location
tab <- "Sulk_Strategy2/full.tab"

#minimum.well.clone
mwc <- 2

#For public clonality remove:
#tissue.to.remove
ttr <- "naive"
```

Install easypackages manager
```{r Install easypackages, message=FALSE}
if("easypackages" %in% rownames(installed.packages()) == F){
  install.packages("easypackages")
}
```


Load and install necessary packages
```{r Load / install necessary packages, message=FALSE}

library(easypackages)
pk <- c("Biostrings","yarrr", "tuple","dplyr","data.table","searchable",
        "tidyverse","openxlsx","stringdist", "gtools","plotly")
pki <- pk[!pk %in% rownames(installed.packages())]
easypackages::install_packages(pki)
libraries(pk)
```

Load raw table from changeO: makeDB.py

```{r Load raw table from changeO: makeDB.py, message=FALSE}
full.tab <- read_delim(file = tab, delim = "\t", col_names = T)
```

Number of total sequences obtained from sequencing:
```{r Number of sequences detected:, message=FALSE}
print(nrow(full.tab))
```

Columns in the database:
```{r Processed by makeDB.py}
colnames(full.tab)
full.tab <- as.data.frame(full.tab)
#saveRDS(full.tab, "Sulk_Results/full.tab.raw.rds")
```

Check and eliminate out_of_frame/unfunctional sequences and special characters
```{r, Check sequences}
trimmed.tab <- full.tab %>% 
  filter(IN_FRAME == TRUE)  %>%
  filter(FUNCTIONAL == TRUE)

#Removes Junctions with non A-Z characters
if(length(grep("[!@#$%^&*(),.?\":{}|<>]", trimmed.tab$JUNCTION)) != 0){
  torm <- grep("[!@#$%^&*(),.?\":{}|<>]", trimmed.tab$JUNCTION, invert = T)
  trimmed.tab <- trimmed.tab[torm,]
  }
```

Translate nt JUNCTION to aa using biostrings
```{r, Translate nt JUNCTION to aa using biostrings}
#Usess biostrings to translate DNA JUNCTIONS into AA
dss <- DNAStringSet(trimmed.tab$JUNCTION)

ass <- translate(dss, genetic.code = GENETIC_CODE, if.fuzzy.codon =  "error")

trimmed.tab$JUNCTION_AA <- as.character(ass)
trimmed.tab$JUNCTION_AA_LENGTH  <- width(ass)
```

Extract extra info from ID
```{r Extract metadata from sequence ID}
#EXTRACT EXTRA METADATA INFO FROM ID
trimmed.tab <- trimmed.tab %>% 
  mutate(BIO.ID = "NA") %>% 
  mutate(PLATE.NUMBER = gsub("^.*(P[0-9]{2}).*", "\\1", trimmed.tab$SEQUENCE_ID))
  
trimmed.tab <- trimmed.tab %>% 
  mutate(PLATE.WELL = gsub("^.*P[0-9]{2}([A-Z][0-9]{2}).*", "\\1", trimmed.tab$SEQUENCE_ID))

trimmed.tab <- trimmed.tab %>% 
  mutate(PLATE_ID = paste0(trimmed.tab$PLATE.NUMBER, trimmed.tab$PLATE.WELL)) %>% 
  mutate(READ.COUNTS = as.numeric(gsub("^.*-","", trimmed.tab$SEQUENCE_ID)))
```

Add Metadata information of wells
```{r Add metadata, include=F, cache=TRUE, error=TRUE}
##ADDS STRAIN_MOUSE_TISSUE ID
##ADD METADATA INFO
trimmed.tab[grep("P0[1][A-Z]0[12]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M1_PP"
trimmed.tab[grep("P0[2][A-Z]0[12]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M4_PP"
trimmed.tab[grep("P0[3][A-Z]0[12]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M1_MLN"
trimmed.tab[grep("P0[4][A-Z]0[12]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M1_naive"
trimmed.tab[grep("P0[5][A-Z]0[12]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M1_PP"
trimmed.tab[grep("P0[6][A-Z]0[12]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M4_PP"
trimmed.tab[grep("P0[7][A-Z]0[12]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M1_MLN"
trimmed.tab[grep("P0[8][A-Z]0[12]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M1_naive"
trimmed.tab[grep("P0[9][A-Z]0[12]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "MM12_M2_MLN"
trimmed.tab[grep("P[1][0][A-Z]0[12]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "MM12_M5_PP"
trimmed.tab[grep("P[1][1][A-Z]0[12]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "MM12_M2_MLN"

trimmed.tab[grep("P0[1][A-Z]0[34]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M1_MLN"
trimmed.tab[grep("P0[2][A-Z]0[34]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M4_MLN"
trimmed.tab[grep("P0[3][A-Z]0[34]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M2_MLN"
trimmed.tab[grep("P0[4][A-Z]0[34]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M1_naive"
trimmed.tab[grep("P0[5][A-Z]0[34]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M1_MLN"
trimmed.tab[grep("P0[6][A-Z]0[34]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M4_MLN"
trimmed.tab[grep("P0[7][A-Z]0[34]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M2_MLN"
trimmed.tab[grep("P0[8][A-Z]0[34]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M1_naive"
trimmed.tab[grep("P0[9][A-Z]0[34]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "MM12_M2_MLN"
trimmed.tab[grep("P[1][0][A-Z]0[34]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "MM12_M6_PP"
trimmed.tab[grep("P[1][1][A-Z]0[34]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "MM12_M5_MLN"

trimmed.tab[grep("P0[1][A-Z]0[56]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M2_PP"
trimmed.tab[grep("P0[2][A-Z]0[56]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M5_PP"
trimmed.tab[grep("P0[3][A-Z]0[56]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M3_MLN"
trimmed.tab[grep("P0[4][A-Z]0[56]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M2_naive"
trimmed.tab[grep("P0[5][A-Z]0[56]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M2_PP"
trimmed.tab[grep("P0[6][A-Z]0[56]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M5_PP"
trimmed.tab[grep("P0[7][A-Z]0[56]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M3_MLN"
trimmed.tab[grep("P0[8][A-Z]0[56]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M2_naive"
trimmed.tab[grep("P0[9][A-Z]0[56]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "MM12_M3_MLN"
trimmed.tab[grep("P[1][0][A-Z]0[56]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "MM12_M7_PP"
trimmed.tab[grep("P[1][1][A-Z]0[56]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "MM12_M6_MLN"

trimmed.tab[grep("P0[1][A-Z]0[78]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M2_MLN"
trimmed.tab[grep("P0[2][A-Z]0[78]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M5_MLN"
trimmed.tab[grep("P0[3][A-Z]0[78]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M4_MLN"
trimmed.tab[grep("P0[4][A-Z]0[78]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M2_naive"
trimmed.tab[grep("P0[5][A-Z]0[78]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M2_MLN"
trimmed.tab[grep("P0[6][A-Z]0[78]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M5_MLN"
trimmed.tab[grep("P0[7][A-Z]0[78]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M4_MLN"
trimmed.tab[grep("P0[8][A-Z]0[78]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M2_naive"
trimmed.tab[grep("P0[9][A-Z]0[78]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "MM12_M3_MLN"
trimmed.tab[grep("P[1][0][A-Z]0[78]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "MM12_M1_MLN"
trimmed.tab[grep("P[1][1][A-Z]0[78]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "MM12_M6_MLN"

trimmed.tab[grep("P0[1][A-Z]0[9]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M3_PP"
trimmed.tab[grep("P0[2][A-Z]0[9]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M6_PP"
trimmed.tab[grep("P0[3][A-Z]0[9]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M5_MLN"
trimmed.tab[grep("P0[4][A-Z]0[9]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M3_naive"
trimmed.tab[grep("P0[5][A-Z]0[9]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M3_PP"
trimmed.tab[grep("P0[6][A-Z]0[9]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M6_PP"
trimmed.tab[grep("P0[7][A-Z]0[9]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M5_MLN"
trimmed.tab[grep("P0[8][A-Z]0[9]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M3_naive"
trimmed.tab[grep("P0[9][A-Z]0[9]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "MM12_M4_MLN"
trimmed.tab[grep("P[1][0][A-Z]0[9]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "MM12_M1_MLN"
trimmed.tab[grep("P[1][1][A-Z]0[9]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "MM12_M7_MLN"

trimmed.tab[grep("P0[1][A-Z][1][0]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M3_PP"
trimmed.tab[grep("P0[2][A-Z][1][0]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M6_PP"
trimmed.tab[grep("P0[3][A-Z][1][0]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M5_MLN"
trimmed.tab[grep("P0[4][A-Z][1][0]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M3_naive"
trimmed.tab[grep("P0[5][A-Z][1][0]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M3_PP"
trimmed.tab[grep("P0[6][A-Z][1][0]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M6_PP"
trimmed.tab[grep("P0[7][A-Z][1][0]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M5_MLN"
trimmed.tab[grep("P0[8][A-Z][1][0]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M3_naive"
trimmed.tab[grep("P0[9][A-Z][1][0]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "MM12_M4_MLN"
trimmed.tab[grep("P[1][0][A-Z][1][0]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "MM12_M1_MLN"
trimmed.tab[grep("P[1][1][A-Z][1][0]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "MM12_M7_MLN"

trimmed.tab[grep("P0[1][A-Z][1][12]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M3_MLN"
trimmed.tab[grep("P0[2][A-Z][1][12]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M6_MLN"
trimmed.tab[grep("P0[3][A-Z][1][12]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M6_MLN"
trimmed.tab[grep("P0[4][A-Z][1][12]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "SPF_M3_naive"
trimmed.tab[grep("P0[5][A-Z][1][12]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M3_MLN"
trimmed.tab[grep("P0[7][A-Z][1][12]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M6_MLN"
trimmed.tab[grep("P0[8][A-Z][1][12]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "GF_M3_naive"
trimmed.tab[grep("P0[9][A-Z][1][12]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "MM12_M4_MLN"
trimmed.tab[grep("P[1][1][A-Z][1][12]",trimmed.tab$PLATE_ID, value = F),]$BIO.ID <- "MM12_M7_MLN"
```

Extract more metadata info from BIO.ID
```{r Extract more metadatainfo from BIO.ID}
#REMOVE BARCODES FROM EMPTY WELLSS
trimmed.tab <- trimmed.tab %>% filter(BIO.ID != "NA")

trimmed.tab <- trimmed.tab %>%
  mutate(STRAIN =  gsub("^([GF12MLNSP]{2,4})_.*","\\1", trimmed.tab$BIO.ID),
         MOUSE =  gsub("^.*_(M[1-7])_.*","\\1",trimmed.tab$BIO.ID),
         TISSUE =  gsub("^.*_(.*$)","\\1",trimmed.tab$BIO.ID))

#Real Mouse info fixed
trimmed.tab <- trimmed.tab %>% mutate(MOUSE =  paste0(trimmed.tab$STRAIN, "_" , trimmed.tab$MOUSE))
```

Splits each dataset by biological samples and tissue (Bio.ID)
```{r Splits each dataset by biological samples and tissue (BIO.ID)}
#Splits dataset by each biological sample and its tissues
trimmed.tab.splited <- split(trimmed.tab, factor(trimmed.tab$BIO.ID))

#Save each splited file in another file
for(i in 1:length(trimmed.tab.splited)){write.table(x = trimmed.tab.splited[[i]], file = sprintf("Sulk_Strategy2/Databases/%s.tab", names(trimmed.tab.splited[i])), sep = "\t", row.names = F, col.names = T)}
```


ChangeO is used in each splited file

DefineClones.py Version
```{bash DefineClones.py Version}
DefineClones.py --version
```
```{bash}
ls
```


```{bash  use DefineClones, eval = FALSE}
DefineClones.py -d final.db.nonnaive.txt --outdir "../Runs/4nt/" --model ham --dist 4 --norm none
```

```{r load DefineClones output}
#DefineClones.py from changeO was used in each database with the nucleotide distance (4nt)
path <- "Sulk_Strategy2/Runs/4nt/"

files <- list.files(path)

files
```

Clonotype analysis per well

```{r Process ChangeO DefineClones output}
#All files are loaded here in a single list
databases <- lapply(sprintf("%s%s", path , files), fread, header = T, sep = "\t")

#Adds mouse name to each df
names(databases) <- gsub("_clone.*", "" , files)

#Merge all tables into the same df
databases <- bind_rows(databases)

#Create unique IDs for each sample 
databases <- databases %>% mutate(CLONE_ID = paste0(databases$BIO.ID, "_", databases$CLONE))
```

Number of detected sequences before single-well collapsing
```{r Number of sequences before}
print(nrow(databases))
```

```{r Well collapsing}
#source("Functions/sulk.fun.R")

mf <- function(x){
  y <- x[!duplicated(x$CLONE_ID), ]
  read.c <- aggregate(x[["READ.COUNTS"]], by=list(CLONE_ID=x[["CLONE_ID"]]), FUN=sum)
  idx <- match(y[["CLONE_ID"]], read.c[["CLONE_ID"]])
  y[["READ.COUNTS.SUM"]] <- read.c[idx,][["x"]]
  return(y)
}

wellcol <- function(databases=databases){
  clone.temp <- split(databases, f = databases$PLATE_ID)
  clone.temp2 <- lapply(clone.temp, FUN = mf)
  final.db <- bind_rows(clone.temp2)
  return(final.db)
}


final.db <- wellcol(databases = databases)
```

Number of detected sequences after single-well collapsing
```{r Number of sequences after collapsing}
print(nrow(final.db))
```

```{r Plot distribution of read number per sequence}
num <- log10(sort(final.db$READ.COUNTS.SUM))

f <- list(
  family = "Courier New, monospace",
  size = 18,
  color = "#7f7f7f"
)

x <- list(
  title = "Sorted by Ascending",
  titlefont = f
)
y <- list(
  title = "log10 (READ.COUNTS.SUM)",
  titlefont = f
)

p <- plot_ly(x = ~1:nrow(final.db)) 
p <- p %>% add_lines(y = num, name = "vh")
p <- p %>% layout(xaxis = x, yaxis = y)

p
```

Number of detected sequences after single-well collapsing and removing low read counts

```{r Remove bad quality sequences}
 final.db <- final.db %>% filter(READ.COUNTS.SUM >= mrn)
```

```{r Number of sequences after removing low read counts}
print(nrow(final.db))
```

Plot number of sequences per well in each group.tissue
```{r Plot number of sequences per well}
#CLone.Wells = cw
#Count the number of clones per well
cw <- table(final.db$CLONE_ID)
# set up cut-off values 
breaks <- c(1, 2 , 3, 4, 7, 13, 33)
# specify interval/bin labels
tags <- c("[1-2)", "[2-3)", "[3-4)","[4-7)","[7-13)", "[13-33)")
# put values into bins
group_tags <- cut(cw, 
                  breaks=breaks, 
                  include.lowest=TRUE, 
                  right=FALSE, 
                  labels=tags)
# inspect bins
summary(group_tags)
#Extract group names for barplot
names(group_tags) <- gsub("([A-Z0-9]{1,10}_[A-Z0-9]{1,10}_[A-Za-z]{1,10})_[0-9]{1,10}","\\1", names(cw))
#Table each clone present in n number of wells


bar.count <- table(names(group_tags), group_tags)
bar.count <- as.matrix(as.data.frame.matrix(bar.count))


#Normalize each value it to 1
bar.count <- bar.count/rowSums(bar.count)
#Specify bargraph order
order <- c(grep('SPF', rownames(bar.count)),
           grep('GF', rownames(bar.count)),
           grep('MM12', rownames(bar.count)))

ordered <- rownames(bar.count)[order]

idx <- grep('naive', ordered)

order <- match(c(ordered[-idx], ordered[idx]),rownames(bar.count))

#Create colors for bars
library(scales)
colors <- hue_pal()(length(colnames(bar.count)))

#Change order
y <- t(bar.count[order,])
# Increase margin size
#Plot bar 

par(mar=c(5,5,1,5)); barplot(y*100, ylab = 'Percentage of sequences in n wells', 
        cex.names=.4, col = colors, horiz = F, las=2, legend.text=TRUE,
        args.legend=list(
          x=ncol(y) + 17.4,
          y=100,
          bty = "n"
        )
)

#Remap 
write.xlsx(y, "Sulk_Results/barplot.table.xlsx", row.names = T)
```

Using a weight corresponding to the clone*well size:
```{r above weighted}
library(GDAtools)
weight <- plyr::revalue(group_tags, c("[1-2)" = 1,
                            "[2-3)" = 2,
                            "[3-4)" = 3,
                            "[4-7)" = 5,
                            "[7-13)" = 10, 
                            "[13-33)" = 23))

y2 <- prop.wtable(names(group_tags), group_tags, w = weight, dir=1, na = F , mar = F)

#Specify bargraph order
order <- c(grep('SPF', rownames(y2)),
           grep('GF', rownames(y2)),
           grep('MM12', rownames(y2)))

ordered <- rownames(y2)[order]

idx <- grep('naive', ordered)

order <- match(c(ordered[-idx], ordered[idx]),rownames(y2))

#Create colors for bars
library(scales)
colors <- hue_pal()(length(colnames(y2)))

#Change order
y <- t(y2[order,])
# Increase margin size
#Plot bar 

par(mar=c(5,5,1,5)); barplot(y, ylab = 'Percentage of sequences in n wells', 
        cex.names=.4, col = colors, horiz = F, las=2, legend.text=TRUE,
        args.legend=list(
          x=ncol(y) + 17.4,
          y=max(colSums(y)),
          bty = "n"
        )
)



```

Plot number of sequences per well:

```{r Sequences per well total}
cells.wells <- table(final.db$PLATE_ID)
names(cells.wells) <- gsub("^(P[0-9]{2})([A-Z][0-9]{2})","\\1_\\2", names(cells.wells))
cells.wells <- cells.wells[mixedorder(names(cells.wells))]
table(final.db$PLATE_ID, final.db$TISSUE)

x <- list(
  title = "Well Bins",
  titlefont = f
)
y <- list(
  title = "Clone.Wells",
  titlefont = f
)

fig <- plot_ly(x = cells.wells, type = "histogram") %>% layout(xaxis = x, yaxis = y)

fig
```


```{r histo}
df0 <- as.data.frame.matrix(table(final.db$PLATE_ID, final.db$TISSUE))
df0 <- as.data.frame(df0)
df0$TISSUE <- 0

for(i in 1:nrow(df0)){ df0$TISSUE[i] <- colnames(df0)[df0[i,] != 0]}


df <- data.frame(index=1:length(cells.wells), wells=cells.wells, sample = df0$TISSUE)

p <- ggplot(data=df, aes(x=wells.Var1, y=wells.Freq, fill = sample)) +
    geom_bar(stat="identity")

fig <- ggplotly(p)

fig
```

Plot number of sequences per well distribution
```{r Pirate Plot of Cells per wells}
df <- as.data.frame(cells.wells)

TISSUE <- final.db[match(as.character(gsub("_", "",as.character(df$Var1))), final.db$PLATE_ID),]$TISSUE
STRAIN <- final.db[match(as.character(gsub("_", "",as.character(df$Var1))), final.db$PLATE_ID),]$STRAIN
BIO <- final.db[match(as.character(gsub("_", "",as.character(df$Var1))), final.db$PLATE_ID),]$BIO.ID

df <- df %>% mutate(STRAIN = STRAIN)
df <- df %>% mutate(TISSUE = TISSUE)
df <- df %>% mutate(BIO = BIO)
df[grep( "naive", df$BIO),]$STRAIN <- "Naive"

df$STRAIN <- factor(df$STRAIN, levels =c("SPF", "GF", "MM12", "Naive"))

write.xlsx(df,"Sulk_Results/pirateplot.table.xlsx")

pirateplot(formula = Freq ~ STRAIN,
           data = df,ylim = c(0,100),
           inf.method	= "iqr",
           main = "Clones Number per Well", xlab = "Groups", ylab = "Number" ,
           theme = 2)



```

```{r}
#saveRDS(final.db, "Sulk_Results/final.db.rds")
```

```{r}
final.db <- final.db %>% dplyr::mutate(VGENE = str_extract(final.db$V_CALL, "IGHV[0-9]{1,2}-[0-9]{1,2}.*"))
final.db <- final.db %>% dplyr::mutate(JGENE = str_extract(final.db$J_CALL, "IGHJ[0-9]{1,2}.*"))
```

```{r}
final.db$VGENE <- gsub("\\*.*","", final.db$VGENE)
final.db$JGENE <- gsub("\\*.*","", final.db$JGENE)
```


```{r}
dbV1_47 <- final.db %>% filter (VGENE == "IGHV1-47")
```
  
```{r}
dist147 <- stringdist(c(dbV1_47$JUNCTION_AA), "CARGSNYDYAMDYW", method = "lv")

prop.dist147 <- data.frame(lv = dist147)

hist <- prop.dist147 %>% dplyr::group_by(lv) %>% summarise(n())

hist$`n()` <- hist$`n()`/sum(hist$`n()`)

p <- ggplot(data=hist, aes(x=lv, y=`n()`)) +
    geom_bar(stat="identity")

fig <- ggplotly(p)

fig
```




```{r Public Clone analysis}
##################################################################
#   ___       _     _ _          _               _           _     
#  / _ \_   _| |__ | (_) ___    /_\  _ __   __ _| |_   _ ___(_)___ 
# / /_)/ | | | '_ \| | |/ __|  //_\\| '_ \ / _` | | | | / __| / __|
#/ ___/| |_| | |_) | | | (__  /  _  \ | | | (_| | | |_| \__ \ \__ \
#\/     \__,_|_.__/|_|_|\___| \_/ \_/_| |_|\__,_|_|\__, |___/_|___/
#                                                  |___/           

##################################################################
```

Remove naive samples</h5></font></strong></div>
```{r filter database}
final.db.nonnaive <- final.db %>% filter(TISSUE != ttr)
print("Number of sequences without naive:")
print(nrow(final.db.nonnaive))

```

```{r AA JUNCTION LENGTH}
final.db.nonnaive <- final.db.nonnaive %>% mutate(JUNCTION_AA_LENGTH = nchar(final.db.nonnaive$JUNCTION_AA))
#write final database for clonotype assignment
#openxlsx::write.xlsx(final.db.nonnaive, "Sulk_Results/final.db.nonnaive.xlsx")
```

```{r}
View(final.db.nonnaive)
```

```{r}
library(dplyr)
```


```{r}
final.db.nonnaive %>% dplyr::group_by(JUNCTION_AA, VGENE, JGENE, PLATE_ID) %>% dplyr::summarise(number = n())  %>% arrange(desc(number))
```



```{r}
Clonality(File = "final.db.nonnaive.xlsx", NewF = "final.db.nonnaive.clonality.xlsx", Vgene_Column = "V_CALL", Jgene_Column = "J_CALL", CDR3_Column = "JUNCTION_AA", Length_Column = "JUNCTION_AA_LENGTH", Mismatch = 0)
```



```{r}
sulk.db <- read_xlsx("final.db.nonnaive.clonality.xlsx")
```


```{r}
sulk.db <- readr::read_delim("DataGen/output/CLONE_4nt_Changeo.tab", delim = "\t")
```



Sorting clones and groupping based on CDR3_V_J_well

each unique sequence can only exist once within each well:

```{r}
sulk.db %>% dplyr::group_by(JUNCTION_AA, VGENE, JGENE, PLATE_ID) %>% dplyr::summarise(number = n())  %>% arrange(desc(number))
```



```{r}
sulk.db$aa_clone <- paste0(sulk.db$JUNCTION_AA,"_",sulk.db$CLONE)

sulk.db.kar <- sulk.db %>% group_by(MOUSE, CLONE) %>% summarise(Size = n(), JUNCTIONS = toString(JUNCTION_AA), VCALL = toString(V_CALL),VGENE = toString(VGENE), JGENE = toString(JGENE)) %>% arrange(desc(Size))
head(sulk.db.kar)
sulk.db.kar
```

```{r}
View(sulk.db.kar)
sulk.db.kar <- sulk.db.kar %>% filter (Size > 1)
nrow(sulk.db.kar)
head(sulk.db.kar)

sulk.db.kar$CLONE_MOUSE <- paste0(sulk.db.kar$CLONE,"_",sulk.db.kar$MOUSE)
head(sulk.db.kar)
# split.sulk <- split(sulk.db.kar, f = sulk.db.kar$CLONE)
# head(split.sulk,1)
# split.sulk <- split.sulk[names(nshar)]
# split.sulk
link_sel <- readr::read_delim("4nt_link.selction.txt", delim = " ", col_names = F)

for(i in 1:nrow(link_sel)){
idx1 <- match(link_sel[i,][1], sulk.db.kar$CLONE_MOUSE)  
idx2 <- match(link_sel[i,][2], sulk.db.kar$CLONE_MOUSE)  
String1 <- sulk.db.kar$JUNCTIONS[idx1]
String2 <- sulk.db.kar$JUNCTIONS[idx2]
 S1 <- gsub(" ", "", stri_list2matrix(sapply(String1,  strsplit,',')))  
 S2 <- gsub(" ", "", stri_list2matrix(sapply(String2,  strsplit,',')))  
 
 if(all(is.na(match(S1,S2))) == FALSE)
 {
  print("yes") 
 }else(
   print("no")
 )
 
} 
 
for(i in 1:nrow(link_sel)){
idx1 <- match(link_sel[i,][1], sulk.db.kar$CLONE_MOUSE)  
idx2 <- match(link_sel[i,][2], sulk.db.kar$CLONE_MOUSE)  
String1 <- sulk.db.kar$JUNCTIONS[idx1]
String2 <- sulk.db.kar$JUNCTIONS[idx2]

StringV1 <- sulk.db.kar$VGENE[idx1]
StringV2 <- sulk.db.kar$VGENE[idx2]
S1 <- gsub(" ", "", stri_list2matrix(sapply(String1,  strsplit,',')))  
S2 <- gsub(" ", "", stri_list2matrix(sapply(String2,  strsplit,',')))  
 
 
if( length(grep("IGHV1-47", StringV1 )) > 0){
  match <- Biostrings::matchPattern("ARGSNY", String1 , max.mismatch = 1, min.mismatch = 1)
  if( length(match)  > 1 ){
    print("red")
  } else{print("black")}
} else{print("black")}
 
 
 
 
  
  
}



for(i in 1:nrow(link_sel)){
idx1 <- match(link_sel[i,][1], sulk.db.kar$CLONE_MOUSE)  
idx2 <- match(link_sel[i,][2], sulk.db.kar$CLONE_MOUSE)  
String1 <- sulk.db.kar$JUNCTIONS[idx1]
String2 <- sulk.db.kar$JUNCTIONS[idx2]

StringV1 <- sulk.db.kar$VGENE[idx1]
StringV2 <- sulk.db.kar$VGENE[idx2]
S1 <- gsub(" ", "", stri_list2matrix(sapply(String1,  strsplit,',')))  
S2 <- gsub(" ", "", stri_list2matrix(sapply(String2,  strsplit,',')))  
 
 
if( length(grep("IGHV1-12", StringV1 )) > 0){
  match1 <- Biostrings::matchPattern("AREGFAY", String1 , max.mismatch = 0)
  match2 <- Biostrings::matchPattern("AREGFAY", String2 , max.mismatch = 0)
  if( length(match1) + length(match2)  > 1 ){
    print("red")
  } else{print("black")}
} else{print("black")}
 
 
 
 
  
  
}


```

```{r}

# intersulk <- function(x){
# S1 <- gsub(" ", "", stri_list2matrix(sapply(y[["JUNCTIONS"]][x[1]],  strsplit,',')))
# S2 <- gsub(" ", "", stri_list2matrix(sapply(y[["JUNCTIONS"]][x[2]],  strsplit,',')))
# if(length(intersect(S1, S2)) != 0){
#   print(intersect(S1, S2))
# }}
# 
# n <- 0
# lapply(
# split.sulk, function(y, name){
# print(n) + 1
#   
# x <- t(combn(x = 1:nrow(y), 2))
# apply(x, MARGIN = 1, FUN = intersulk)
# }
# )







```

```{r}
# nshar <- sort(unlist(lapply(split.sulk, nrow)), decreasing = T)
# nshar <- nshar[nshar > 1]
# 
# prep_ov <- function(x){
#   strsplit(x[["JUNCTIONS"]], ",")}
# 
# ov <- lapply(split.sulk, prep_ov)
# 
# ov <- ov[names(nshar)]
# 
# ov$`4487`
# 
# 


```


```{r}
#View(sulk.db %>% filter(CLONE == "8667"))
```

```{r}
sulk.db.kar <- split(sulk.db.kar, f = sulk.db.kar$MOUSE)


CircosOut::circos.karyotype(karyotype = sulk.db.kar, CDR3 = "CLONE", Group = "MOUSE", save = T, file.name = "karyotype_aa0.txt")
```

```{r}
bands <- readr::read_delim("Circos/0aa_Changeo/data/karyotype.txt", delim = " ", col_names = F)
bands <- bands[grep("band", bands$X1, invert = F), ]
head(bands)
```


```{r}
CircosOut::links(bands = bands)
```




Defining the global design for all sequences:

```{r}
sulk.circos <- final.db.nonnaive %>% dplyr::group_by(PLATE_ID) %>% dplyr::summarise(Size = n()) %>% dplyr::arrange(desc(Size))
```

```{r}
library(CircosOut)
sulk.circos$mouse <- final.db.nonnaive$MOUSE[match(sulk.circos$PLATE_ID, final.db.nonnaive$PLATE_ID)]

sulk.circos <- split(sulk.circos, f = sulk.circos$mouse)

circos.karyotype(karyotype = sulk.circos, CDR3 = "PLATE_ID", Group = "mouse", save = T, file.name = "sulk.karyotype.txt")







```


```{r Access inter-sample clonality}


nrow(final.db.nonnaive[duplicated(final.db.nonnaive$CLONE_ID),])

names.idx <- names(table(final.db.nonnaive$CLONE_ID))[table(final.db.nonnaive$CLONE_ID) >= mwc ]
temp <- final.db.nonnaive[final.db.nonnaive$CLONE_ID %in% names.idx,]

table(sort(table(final.db.nonnaive$CLONE_ID)))

write.xlsx(final.db.nonnaive, "final.db.nonnaive.xlsx")


Clonality(File = "final.db.nonnaive.xlsx", NewF = "clonality/final.db.nonnaive.clonality.xlsx", Vgene_Column = "V_CALL",
          Jgene_Column = "J_CALL", CDR3_Column = "JUNCTION_AA", Length_Column = "JUNCTION_AA_LENGTH", Mismatch = 0)

```


```{r Circos conf files creation}

```

